### PR TITLE
Objectify NeoMutt

### PR DESCRIPTION
**Pull request by nabijaczleweli**
* **What does this PR do?**
struct NeoMutt *NeoMutt; -> struct NeoMutt NeoMutt;

There's no reason to have four pointer chases when you can have three.

First three commits automatic, as described in the messages.

Hoping CI has an allyesconfig configuration and will catch anything my minimal ./configure has missed.

### CODE REVIEW

**Comment by flatcap** [comment]
> Hoping CI has an allyesconfig configuration

Yes
"Build and Test" does default, min and max builds.
 "ASAN" does a max build and test under the [Address Sanitizer](https://neomutt.org/dev/analysis/asan)

**Comment by flatcap** [comment]
Files changed 201
```diff
 diff
+1,279
-1,289
```

Hmm... this is going to make rebasing the other PRs interesting.

**Comment by flatcap** [comment]
Yeah, given that it's all `sed`-able, it shouldn't be a problem.

A technique I often use, is to turn the PR branch into patches (`git format-patch`),
then manipulate the patches (vim-fu, sed, etc), then reapply them.

**Comment by flatcap** [comment]
I've just noticed your branch name!  `NeoMutt->NeoMutt->`
I'm amazed all the tooling between us works!

**Comment by flatcap** [comment]
Time to tick the box!

**Comment by gahr** [comment]
You are now using `NeoMutt.sub` as an indication that `NeoMutt` was init'd. In some places, you use `NeoMutt.notify` for that, because the code uses that member. But the code happens to also use `NeoMutt.sub`. I think it's a bit implicit. How about an explicit `bool initialized` member?

**Comment by nabijaczleweli** [comment]
yeah, that's definitely sensible

File: core/neomutt.c [hunk 1]
```diff
@@ -34,25 +34,24 @@
...
-struct NeoMutt *NeoMutt = NULL; ///< Global NeoMutt object
+struct NeoMutt NeoMutt; ///< Global NeoMutt object
 
 /**
  * neomutt_new - Create the main NeoMutt object
  * @param cs Config Set
  * @retval ptr New NeoMutt
  */
-struct NeoMutt *neomutt_new(struct ConfigSet *cs)
+struct NeoMutt neomutt_new(struct ConfigSet *cs)
```
Comment (gahr) [inline core/neomutt.c:44]: s/neomutt_new/neomutt_init/ and take a `struct Neomutt*` as argument

Comment (nabijaczleweli) [reply core/neomutt.c]: applied

**Comment by nabijaczleweli** [comment]
aaand CI passes again

**Comment by ossilator** [comment]
you can avoid that enormous churn by declaring the global as
`struct NeoMutt NeoMutt[1];`. that would also keep the code more uniform, and avoid repeated churn if somebody changed it again (e.g., to lib-ify it).

**Comment by nabijaczleweli** [comment]
which also hides `if(NeoMutt)` being broken

**Comment by flatcap** [comment]
@ossilator
That's cute, but the idea was to save time on all the dereferences.

I'm not too worried about churn.
The current users of `NeoMutt` are:

 - 10 `NeoMutt.accounts`
 - 56 `NeoMutt.notify`
 - 1017 `NeoMutt.sub`
 
The config system supports Account- and Mailbox-specific variables.
When we implement these, all the `NeoMutt.sub` will disappear.

**Comment by flatcap** [comment]
> the generated code is the same as with nab's patch

Ah!  Fair enough

**Comment by flatcap** [comment]
Thanks for all the feedback.
I'm happy to consider enormous code changes, but in this case I'm going to say, "**No, thanks**".

- nabijaczleweli's changes reduced the visual clutter and saved one asm instruction each
- ossilator's suggestion cleverly won the same asm saving, without the churn

94% of the `NeoMutt` dereferences are for config lookups.
I'd rather focus on eliminating them.
Doing so would actually restore the pointer derefence, albeit on a different object,
e.g. `account->sub` or `mailbox->sub`.

**Comment by flatcap** [comment]
Archived to https://github.com/neomutt/neomutt-old/tree/neomutt-object

**Pull request closed without merging**